### ISSUE TITLE
Rubygems v3.2.3 bug in command 'gem wrappers regenerate'

### ISSUE THREAD

**Issue by pedrofurtado** [rubygems/rubygems#4204]
I'm having a problem.

My current problem is:

When I install ruby 3.0.0 + rubygem v3.2.3 and execute `gem wrappers regenerate`, the following error occurs in terminal:

```bash
$ gem wrappers regenerate
ERROR:  While executing gem ... (Gem::CommandLineError)
    Unknown command wrappers
```

This issue is related to:

  - [ ] Network problems
  - [ ] Installing a library
  - [ ] Publishing a library
  - [x] The command line `gem`
  - [ ] Other

Here are my current environment details:

```
$ cat /etc/centos-release
CentOS Linux release 8.3.2011
```

```
$ gem env version
3.2.3
```

```
$ gem environment
RubyGems Environment:
  - RUBYGEMS VERSION: 3.2.3
  - RUBY VERSION: 3.0.0 (2020-12-25 patchlevel 0) [x86_64-linux]
  - INSTALLATION DIRECTORY: /usr/local/rvm/gems/ruby-3.0.0
  - USER INSTALLATION DIRECTORY: /root/.local/share/gem/ruby/3.0.0
  - RUBY EXECUTABLE: /usr/local/rvm/rubies/ruby-3.0.0/bin/ruby
  - GIT EXECUTABLE: /bin/git
  - EXECUTABLE DIRECTORY: /usr/local/rvm/gems/ruby-3.0.0/bin
  - SPEC CACHE DIRECTORY: /root/.local/share/gem/specs
  - SYSTEM CONFIGURATION DIRECTORY: /usr/local/rvm/rubies/ruby-3.0.0/etc
  - RUBYGEMS PLATFORMS:
     - ruby
     - x86_64-linux
  - GEM PATHS:
     - /usr/local/rvm/gems/ruby-3.0.0
     - /usr/local/rvm/rubies/ruby-3.0.0/lib/ruby/gems/3.0.0
  - GEM CONFIGURATION:
     - :update_sources => true
     - :verbose => true
     - :backtrace => false
     - :bulk_threshold => 1000
  - REMOTE SOURCES:
     - https://rubygems.org/
  - SHELL PATH:
     - /usr/local/rvm/gems/ruby-3.0.0/bin
     - /usr/local/rvm/gems/ruby-3.0.0@global/bin
     - /usr/local/rvm/rubies/ruby-3.0.0/bin
     - /usr/local/rvm/bin
     - /sbin
     - /bin
     - /usr/sbin
     - /usr/bin
```

```
$ ruby -v
ruby 3.0.0p0 (2020-12-25 revision 95aff21468) [x86_64-linux]
```

I will abide by the [code of conduct](https://github.com/rubygems/rubygems/blob/master/CODE_OF_CONDUCT.md).

**Comment by pedrofurtado** [comment]
A workaround for a while is to downgrade the version to 3.2.2, as below:

```
$ gem update --system 3.2.2
Updating rubygems-update
Fetching rubygems-update-3.2.2.gem
Successfully installed rubygems-update-3.2.2
Parsing documentation for rubygems-update-3.2.2
Installing ri documentation for rubygems-update-3.2.2
Installing darkfish documentation for rubygems-update-3.2.2
Done installing documentation for rubygems-update after 69 seconds
Parsing documentation for rubygems-update-3.2.2
Done installing documentation for rubygems-update after 0 seconds
Installing RubyGems 3.2.2
  Successfully built RubyGem
  Name: bundler
  Version: 2.2.2
  File: bundler-2.2.2.gem
Bundler 2.2.2 installed
RubyGems 3.2.2 installed
Regenerating binstubs
Regenerating plugins
Parsing documentation for rubygems-3.2.2
Installing ri documentation for rubygems-3.2.2



------------------------------------------------------------------------------

RubyGems installed the following executables:
        /usr/local/rvm/rubies/ruby-3.0.0/bin/gem
        /usr/local/rvm/rubies/ruby-3.0.0/bin/bundle

Ruby Interactive (ri) documentation was installed. ri is kind of like man
pages for Ruby libraries. You may access it like this:
  ri Classname
  ri Classname.class_method
  ri Classname#instance_method
If you do not wish to install this documentation in the future, use the
--no-document flag, or set it as the default in your ~/.gemrc file. See
'gem help env' for details.

RubyGems system software updated
```

And then execute the command `gem wrappers regenerate`:

```
$ gem wrappers regenerate
# Nothing appears in terminal, as expected
```

**Comment by deivid-rodriguez** [comment]
Not sure what's going on, I installed the `gem-wrappers` gem on my system using 3.2.3 and the command worked.

**Comment by pedrofurtado** [comment]
@deivid-rodriguez Thanks for feedback! Maybe this can be related: I installed ruby 3.0 + rubygems with RVM 👍

**Comment by pedrofurtado** [comment]
But, @deivid-rodriguez , I don't know how to debug to find root cause 🙈  Is there any "verbose mode" for command `gem wrappers regenerate`?

**Comment by deivid-rodriguez** [comment]
I don't know, I don't really use `RVM` and don't know much about that command. If you can provide me with a `Dockerfile` reproducing the error I can try look into it.

**Comment by pedrofurtado** [comment]
I will provide a minimal Vagrantfile 🤝

**Comment by pedrofurtado** [comment]
@deivid-rodriguez If you have Vagrant+Virtualbox in your PC, you can use this Vagrantfile as example to reproduce the error:

```ruby
Vagrant.configure('2') do |config|
  config.vm.box = 'centos/8'
  config.vm.box_version = '2011.0'

  config.vm.provision 'shell', privileged: true, inline: <<-SHELL
    cat >> /home/vagrant/.bashrc <<EOL
sudo su
source /etc/profile.d/rvm.sh
EOL

    echo "retries=0" >> /etc/yum.conf
    echo "timeout=10000" >> /etc/yum.conf
    yum check-update -y
    yum update -y
    yum -y install epel-release
    yum check-update -y
    yum update -y
    yum install -y dnf-plugins-core
    yum config-manager --set-enabled powertools
    yum install -y git cronie vim curl libcurl-devel curl-devel tar wget lynx sed zip unzip openssh-server htop gpg gnupg2 make gcc epel-release yum-utils
    yum groupinstall -y 'Development Tools'
    systemctl enable sshd
    systemctl start sshd

    command curl -sSL https://rvm.io/mpapis.asc | gpg2 --import -
    command curl -sSL https://rvm.io/pkuczynski.asc | gpg2 --import -
    curl -sSL https://get.rvm.io | bash -s stable
    source /etc/profile.d/rvm.sh
    rvm get head
    rvm install 3.0.0
  SHELL
end
```

In a terminal, execute these commands:

```bash
cd folder/where/Vagrantfile/is/
vagrant up # Will create the VM and execute the failed "gem wrappers regenerate"
vagrant ssh # Will connect SSH inside VM
```

**Comment by pedrofurtado** [comment]
@deivid-rodriguez Could you reproduce the error? 🤝

**Comment by wintersolutions** [comment]
This seems not to be a issue in `rubygems`. I believe the issue is in the project https://github.com/rvm/gem-wrappers which seems to act as a plugin to `rubygems` and provides the `gem wrappers regenerate` command.

**Comment by hsbt** [comment]
It seems the 3rd party issue.

**Comment by deivid-rodriguez** [comment]
@pedrofurtado I didn't try to repro the error yet, but I will try to find some time. If you could prepare a `Dockerfile` though, it'd be easier for me :+1:.

**Comment by pedrofurtado** [comment]
Thanks for explanation @wintersolutions ! I will provide this information to RVM team 🤝 

Anyway, If you could help us @deivid-rodriguez , it will be awesome for debugging (using your expertise in rubygems codebase 🔍 🕵️).

I will create asap a Dockerfile 👍

**Comment by pedrofurtado** [comment]
@hsbt Could you please reopen the issue for a while? 🤝

**Comment by pedrofurtado** [comment]
@deivid-rodriguez 

Dockerfile to reproduce the error:

```Dockerfile
FROM centos:8

RUN echo "source /etc/profile.d/rvm.sh" >> /root/.bashrc

RUN echo "retries=0" >> /etc/yum.conf
RUN echo "timeout=10000" >> /etc/yum.conf
RUN yum check-update -y || true
RUN yum update -y
RUN yum -y install epel-release
RUN yum check-update -y || true
RUN yum update -y
RUN yum install -y dnf-plugins-core
RUN yum config-manager --set-enabled powertools
RUN yum install -y git cronie vim curl libcurl-devel curl-devel tar wget lynx sed zip
unzip openssh-server htop gpg gnupg2 make gcc epel-release yum-utils
RUN yum groupinstall -y 'Development Tools'

RUN command curl -sSL https://rvm.io/mpapis.asc | gpg2 --import -
RUN command curl -sSL https://rvm.io/pkuczynski.asc | gpg2 --import -
RUN curl -sSL https://get.rvm.io | bash -s stable
RUN bash -l -c "source /etc/profile.d/rvm.sh && rvm get head"

CMD while :; do echo 'Sleeping ...'; sleep 10; done;
``` 

And then, the commands required for it:

```bash
cd folder/where/Dockerfile/is/
docker build -t mydebugimg .
docker container run -d --name mydebug mydebugimg
docker container exec -it mydebug /bin/bash

# Inside container, run this command:
rvm install 3.0.0
```

**Comment by deivid-rodriguez** [comment]
@pedrofurtado This issue was probably caused by a recent rubygems release, but whatever broke might be completely unsupported and not an actual issue in rubygems, we don't know what happened yet :man_shrugging:. Even if unsupported, we might still choose to ship a workaround, but we don't need to reopen the issue unless we confirm this is a problem in rubygems.

**Comment by deivid-rodriguez** [comment]
Thanks, the `Dockerfile` is helpful :+1:.

**Comment by pedrofurtado** [comment]
Hmmm, thank you so much for your feedback @deivid-rodriguez 🤝 ! Now it is more clear the situation 🤔

**Comment by wintersolutions** [comment]
I tried to fix this last evening but the tests are almost killing my old laptop I have on the road.

I believe the error happens in https://github.com/rvm/rvm/blob/master/scripts/functions/installer#L1420

> __rvm_with "${__gem_home##*/}" gem wrappers regenerate

The error can be solved with a call to `gem pristine gem-wrappers --only-plugins` which is available since `rubygems` v 3.2.

So I tried the following:

> __rvm_with "${__gem_home##*/}" (gem pristine gem-wrappers --only-plugins || true) && gem wrappers regenerate

But the tests don't go trough on branches with those changes, they don't go trough on `master` as well tough (laptop). I don't know if its really working, if its the most effective way to do it and how many error messages this will throw on `rubygems` < v3.2 but its a start.

**Comment by deivid-rodriguez** [comment]
Ok, I had a look at this and yeah, @wintersolutions is correct that this is due to the new plugins layout introduction in #3108.

I don't know much about how `rvm` works, but I think something like this reproduces the situation well:

```
GEM_HOME=foo gem install gem-wrappers && GEM_PATH=foo gem wrappers
```

Rubygems should look for plugin files in all `GEM_PATH` locations, not only in `GEM_HOME`. I'll try to fix this.

**Comment by pedrofurtado** [comment]
@deivid-rodriguez 🎉 🤩

**Comment by pedrofurtado** [comment]
Happy new Year for all! 🍻 📅 

Let us know, @deivid-rodriguez , when you have any news about it. I will be glad to help with testing 🤝

**Comment by deivid-rodriguez** [comment]
I had a look yesterday, but it turns out it's a bit more complicated than I thought. I will post an update soon.

**Comment by WillRadi** [comment]
@deivid-rodriguez :
Is there any news? :)

Thanks, man! Happy new year!

**Comment by deivid-rodriguez** [comment]
No, sorry! I couldn't look again yet.

To explain the problem, now rubygems keeps tracks of a list of enabled rubygems plugins in a `plugins` folder inside the ruby installation layout. Only the plugin for the latest version of each gem installed is kept there, so if I have foo-1.0 and foo-1.1, the plugin for foo 1.1 is kept there but if I uninstall foo-1.1, then the plugin for the remaining foo-1.0 will replace it since it's now the highest version installed. And if then I install foo-1.2, then the plugin for foo 1.2 replaces the one for foo-1.0 since it's now the latest version installed.

Keeping track of the above is a bit tricky since it requires to regenerate plugins both after installating and uninstalling gems, but if we have to do the same thing, but for several GEM_PATH locations, it gets even trickier.

When I looked at the issue the first time I recall that `gem install gem-wrappers` was actually failing for the second gemset rvm creates with a "latest version already installed" message, and that's  what was preventing the plugins from being regenerated for the new gemset. Not sure if that's expected or not, but explains why https://github.com/rvm/rvm/issues/5020#issuecomment-752025534 suggestion by @wintersolutions seems to work.

To sum up, I'm not sure when I'll get to this, and I'm not fully sure anymore whether this is something that should be fixed on our side or on rvm side. For now if RVM can offer a workaround for their users in the mean time I would appreciate it.

**Comment by wintersolutions** [comment]
@deivid-rodriguez I'm trying to get this fixed on the `rvm` side https://github.com/rvm/rvm/pull/5027 now. Thanks for the input! I'll keep you updated.

**Comment by pedrofurtado** [comment]
@deivid-rodriguez Feedback: The hotfix provided by @wintersolutions is available inside rvm codebase (https://github.com/rvm/rvm/pull/5027 merged ✅ ). Thanks a lot, @deivid-rodriguez , of your help in debugging with us. 

Now, we just have to wait an official release of rvm to upgrade to ruby 3 in our systems here 🎉 🍻

**Comment by wintersolutions** [comment]
@deivid-rodriguez yep its merged... 

Thanks to everybody involved!

**Comment by deivid-rodriguez** [comment]
Thanks so much @wintersolutions and @pedrofurtado for stepping in to help here. I'll try to make the next release smoother for `RVM` users.
